## C5 P11

### (Computer Problem 5.4.1 acd) Use Adaptive Trapezoid Quadrature to approximate the integrals within $0.5 x 10^{-8}$. Report the answer with eight correct decimal places and the number of subintervals required.


### (a)

$$ \int_{0}^{4} \frac{x}{\sqrt{x^2 + 9}} \; dx = 2$$

In [1]:
# import needed libraries
import numpy as np

def trap(f, a, b):
    s = (f(a)+f(b))*(b-a)/2.0
    return s

def adapquad(f, a0, b0, tol):
    sum = 0.0
    n = 0
    m = np.array([])
    alist = [a0]
    blist = [b0]
    tols = [tol]
    apps = [trap(f,a0,b0)]
    
    while len(alist) > 0:
        if n > 15:
            quit()
        a = alist.pop()
        b = blist.pop()
        oldtol = tols.pop()
        oldapp = apps.pop()
        c = (a+b)/2.0
        leftapp = trap(f,a,c)
        rightapp = trap(f,c,b)
        errest = oldapp - (leftapp + rightapp)
        if abs(errest) < 3.*oldtol:
            sum  += leftapp+rightapp
        else:
# divide into two intervals
            alist.append(a)
            blist.append(c)
            alist.append(c)
            blist.append(b)
            tols.append(oldtol/2.)
            tols.append(oldtol/2.)
            apps.append(leftapp)
            apps.append(rightapp)
        m = np.append(m,c)
    return [sum, m]

def f(x):
    return x/np.sqrt(np.power(x,2) + 9)

a = 0
b = 4
tol = 0.5*np.power(10.0,-8.0)
appr = adapquad(f, a, b, tol)

print('approximate value = ', appr[0])
print('number of subintervals = ', appr[1].size-1)

approximate value =  1.9999999975
number of subintervals =  12604


$$ \int_{0}^{4} \frac{x}{\sqrt{x^2 + 9}} \; dx \approx 1.9999999975$$

The number of subintervals needed is $12604$.

### (c)

$$ \int_{0}^{1} x \, e^x \; dx = 1$$

In [2]:
# import needed libraries
import numpy as np

def trap(f, a, b):
    s = (f(a)+f(b))*(b-a)/2.0
    return s

def adapquad(f, a0, b0, tol):
    sum = 0.0
    n = 0
    m = np.array([])
    alist = [a0]
    blist = [b0]
    tols = [tol]
    apps = [trap(f,a0,b0)]
    
    while len(alist) > 0:
        if n > 15:
            quit()
        a = alist.pop()
        b = blist.pop()
        oldtol = tols.pop()
        oldapp = apps.pop()
        c = (a+b)/2.0
        leftapp = trap(f,a,c)
        rightapp = trap(f,c,b)
        errest = oldapp - (leftapp + rightapp)
        if abs(errest) < 3.*oldtol:
            sum  += leftapp+rightapp
        else:
# divide into two intervals
            alist.append(a)
            blist.append(c)
            alist.append(c)
            blist.append(b)
            tols.append(oldtol/2.)
            tols.append(oldtol/2.)
            apps.append(leftapp)
            apps.append(rightapp)
        m = np.append(m,c)
    return [sum, m]

def f(x):
    return x*np.exp(x)

a = 0
b = 1
tol = 0.5*np.power(10.0,-8.0)
appr = adapquad(f, a, b, tol)

print('approximate value = ', appr[0])
print('number of subintervals = ', appr[1].size-1)

approximate value =  1.00000000269
number of subintervals =  12422


$$ \int_{0}^{4} \frac{x}{\sqrt{x^2 + 9}} \; dx \approx 1.00000000269$$

The number of subintervals needed is $12422$.

### (d)

$$ \int_{1}^{3} x^2 \ln{(x)} \; dx = 6.99862170912$$

In [3]:
# import needed libraries
import numpy as np

def trap(f, a, b):
    s = (f(a)+f(b))*(b-a)/2.0
    return s

def adapquad(f, a0, b0, tol):
    sum = 0.0
    n = 0
    m = np.array([])
    alist = [a0]
    blist = [b0]
    tols = [tol]
    apps = [trap(f,a0,b0)]
    
    while len(alist) > 0:
        if n > 15:
            quit()
        a = alist.pop()
        b = blist.pop()
        oldtol = tols.pop()
        oldapp = apps.pop()
        c = (a+b)/2.0
        leftapp = trap(f,a,c)
        rightapp = trap(f,c,b)
        errest = oldapp - (leftapp + rightapp)
        if abs(errest) < 3.*oldtol:
            sum  += leftapp+rightapp
        else:
# divide into two intervals
            alist.append(a)
            blist.append(c)
            alist.append(c)
            blist.append(b)
            tols.append(oldtol/2.)
            tols.append(oldtol/2.)
            apps.append(leftapp)
            apps.append(rightapp)
        m = np.append(m,c)
    return [sum, m]

def f(x):
    return np.power(x,2.0)*np.log(x)

a = 1
b = 3
tol = 0.5*np.power(10.0,-8.0)
appr = adapquad(f, a, b, tol)

print('approximate value = ', appr[0])
print('number of subintervals = ', appr[1].size-1)

approximate value =  6.99862171179
number of subintervals =  32766


$$ \int_{0}^{4} \frac{x}{\sqrt{x^2 + 9}} \; dx \approx 6.99862171179$$

The number of subintervals needed is $32766$.